src/visualizers/VisualizerGrid.ts

```diff
@@ -0,0 +1,1191 @@
+import {VisualizerExportModule} from '@/visualizers/VisualizerInterface'
+import {VisualizerDefault} from '@/visualizers/VisualizerDefault'
+import {floorSqrt} from '@/shared/math'
+import type {ParamInterface} from '@/shared/Paramable'
+
+/** md
+# Grid Visualizer
+
+[<img src="../../assets/img/Grid/example-grid.png" width="320" 
+style="margin-left: 1em; margin-right: 1em" 
+/>](../../assets/img/Grid/example-grid.png)
+
+This visualizer puts a sequence in a square spiral or in
+rows and allows you to highlight numbers based on various
+properties.
+
+The inspiration for this visualizer is [Ulam's
+spiral](https://en.wikipedia.org/wiki/Ulam_spiral), which puts the
+natural numbers in a square spiral and highlights the primes.  One can
+also highlight properties such a whether a number is abundant or
+polygonal. Several properties can be highlighted at once, in which case
+later properties overcolor earlier ones.
+
+
+## Parameters
+**/
+const BLACK = '#000000'
+const WHITE = '#ffffff'
+
+const RED = '#ff5733'
+const ORANGE = '#ff9d33'
+const YELLOW = '#e6ff33'
+const GREEN = '#14cd33'
+const BLUE = '#3388ff'
+const PURPLE = '#8814cd'
+const CYAN = '#00ffff'
+const MAGENTA = '#ff00ff'
+const VERDANT = '#9BBF30'
+const VIOLET = '#7f00ff'
+const MUSTARD = '#ffdb58'
+const GRAY = '#808080'
+
+const DEFAULT_COLORS = [
+    RED,
+    ORANGE,
+    YELLOW,
+    GREEN,
+    BLUE,
+    PURPLE,
+    CYAN,
+    MAGENTA,
+    VERDANT,
+    VIOLET,
+    MUSTARD,
+    GRAY,
+]
+
+const RAINBOW = [
+    '#FF0000',
+    '#FF6700',
+    '#FFE000',
+    '#51FF00',
+    '#00F4E8',
+    '#00B1FF',
+    '#0055FF',
+    '#C000FF',
+    '#FF00F9',
+    '#FF0080',
+]
+
+const MAXIMUM_ALLOWED_PROPERTIES = 12
+
+enum Preset {
+    Custom,
+    Primes,
+    Abundant_Numbers,
+    Abundant_Numbers_And_Primes,
+    Polygonal_Numbers,
+    Color_By_Last_Digit_1,
+    Color_By_Last_Digit_2,
+}
+
+enum PathType {
+    Spiral,
+    Rows,
+    Rows_Offset,
+}
+
+enum Direction {
+    Right,
+    Left,
+    Up,
+    Down,
+    StartNewRow,
+    None,
+}
+
+enum Property {
+    None,
+    Prime,
+    Negative,
+    Even,
+    Divisible_By_Three,
+    Divisible_By_Four,
+    Divisible_By_Five,
+    Divisible_By_Six,
+    Divisible_By_Seven,
+    Divisible_By_Eight,
+    Ends_With_Zero,
+    Ends_With_One,
+    Ends_With_Two,
+    Ends_With_Three,
+    Ends_With_Four,
+    Ends_With_Five,
+    Ends_With_Six,
+    Ends_With_Seven,
+    Ends_With_Eight,
+    Ends_With_Nine,
+    Triangular_Number,
+    Square_Number,
+    Pentagonal_Number,
+    Hexagonal_Number,
+    Heptagonal_Number,
+    Octagonal_Number,
+    Sum_Of_Two_Squares,
+    Abundant,
+    Perfect,
+    Deficient,
+    Semi_Prime,
+}
+
+enum PropertyVisualization {
+    Fill_Box,
+    Small_Box,
```

Comment:
My newest commit changed the names here, so it is "Property Style" and you choose from "Fill_Box" and "Small_Box" (instead of Property Visualization and you choose from Primary and Secondary).  This felt much more self-explanatory to me once I was writing the docs.

Reply:
Maybe the input label for this in the dialog could just be "Display as:"? With the options "Fill_Box" and "Small_Box", it should be pretty understandable, and would avoid the need to repeat "Property N"... And should that first option actually be "Fill_Cell"? (To keep the language around cells consistent?) For that matter, should the label on "Property 1 Color" be changed to just "Color" (since after all, it is in the Property 1 "section")?

Reply:
Oh, nice idea!  I'm not sure about "Fill_Cell" only because then "Small_Box" is less understandable.  I thought of "Fill_Cell" and "Mark_Cell" but I thought that wouldn't be clear either.  Maybe "Fill_Cell" and "Fill_Center"?  I don't find any of these options super great.

Reply:
Maybe "Fill_Cell" and "Box_In_Cell"? Alternatively, we could change it to "Indicator size:" and have that be a decimal from 0.1 to 1.0, for the fraction of the size of the cell taken up by the (centered square) indicator of this property. That way you _could_ do three overlaid, for a large enough screen and few enough cells... I am just brainstorming, please go with whatever seems reasonable.

Reply:
> Of course if you would like me to take a stab at such fixes, just let me know.

I fixed the missing image file now.  But the warnings I took a stab at and wasn't able to figure out why the histogram would be different (the code all looks the same to me).  If you could take a stab at the warnings, that would be great.  In the most recent commit, I also fixed the small things to do with the params dialog, and I changed a name again:  I replaced Rows_Offset with Rows_Augmented because this is closer to the meaning of that option, now that you explained it to me.

Reply:
Why did my reply end up in this comment chain?  Github must have got confused.  I was responding to the docs warnings comment.

Reply:
> Maybe "Fill_Cell" and "Box_In_Cell"? Alternatively, we could change it to "Indicator size:" and have that be a decimal from 0.1 to 1.0, for the fraction of the size of the cell taken up by the (centered square) indicator of this property. That way you _could_ do three overlaid, for a large enough screen and few enough cells... I am just brainstorming, please go with whatever seems reasonable.

That's a cool idea.  Maybe I'll try that.

Reply:
Ok, I'm going with Fill_Cell and Box_In_Cell rather than adding functionality at this point; would like to get this PR done.

Reply:
done; please take a look

Reply:
Looks like `Fill_Cell` and `Box_In_Cell` were added in https://github.com/numberscope/frontscope/pull/230/commits/bf0ceed0fdad013675d43d859d418b4b6db2ccbb and https://github.com/numberscope/frontscope/pull/230/commits/f23f1284c6ce2320bbc9e669032e5ca11f7ac73e, respectively. Resolving.